lib/TH/THAtomic.c

```diff
@@ -0,0 +1,90 @@
+#include "THAtomic.h"
+
+/*
+  Note: I thank Leon Bottou for his useful comments.
+  Ronan.
+*/
+
+#if defined(USE_MSC_ATOMICS)
+#include <intrin.h>
+#endif
+
+#if !defined(USE_MSC_ATOMICS) && !defined(USE_GCC_ATOMICS) && defined(USE_PTHREAD_ATOMICS)
+#include <pthread.h>
+static pthread_mutex_t ptm = PTHREAD_MUTEX_INITIALIZER;
+#endif
+
+int THAtomicSet(int volatile *a, int newvalue)
+{
+#if defined(USE_MSC_ATOMICS)
+  return _InterlockedExchange((long*)a, newvalue);
+#elif defined(USE_GCC_ATOMICS)
+  return __sync_lock_test_and_set(a, newvalue);
+#else
+  int oldvalue;
+  do {
+    oldvalue = *a;
+  } while (!THAtomicCompareAndSwap(a, oldvalue, newvalue));
+  return oldvalue;
+#endif
+}
+
+int THAtomicGet(int volatile *a)
+{
+  int value;
+  do {
+    value = *a;
+  } while (!THAtomicCompareAndSwap(a, value, value));
+  return value;
+}
+
+int THAtomicAdd(int volatile *a, int value)
+{
+#if defined(USE_MSC_ATOMICS)
+  return _InterlockedExchangeAdd((long*)a, value);
+#elif defined(USE_GCC_ATOMICS)
+  return __sync_fetch_and_add(a, value);
+#else
+  int oldvalue;
+  do {
+    oldvalue = *a;
+  } while (!THAtomicCompareAndSwap(a, oldvalue, (oldvalue + value)));
+  return oldvalue;
+#endif
+}
+
+void THAtomicIncrementRef(int volatile *a)
+{
+  THAtomicAdd(a, 1);
+}
+
+int THAtomicDecrementRef(int volatile *a)
+{
+  return (THAtomicAdd(a, -1) == 1);
+}
+
+int THAtomicCompareAndSwap(int volatile *a, int oldvalue, int newvalue)
+{
+#if defined(USE_MSC_ATOMICS)
+  return (_InterlockedCompareExchange((long*)a, (long)newvalue, (long)oldvalue) == (long)oldvalue);
+#elif defined(USE_GCC_ATOMICS)
+  return __sync_bool_compare_and_swap(a, oldvalue, newvalue);
+#elif defined(USE_PTHREAD_ATOMICS)
+  int ret = 0;
+  pthread_mutex_lock(&ptm);
+  if(*a == oldvalue) {
+    *a = newvalue;
+    ret = 1;
+  }
+  pthread_mutex_unlock(&ptm);
+  return ret;
+#else
+#warning THAtomic is not thread safe
+  if(*a == oldvalue) {
+    *a = newvalue;
```

Comment:
And if you cared about this case at all (probably not, since it's not thread safe :) ), `volatile*` likely doesn't do what you think it does. `volatile` is only a compiler barrier, not a memory barrier. There's no guarantee that any other thread would ever see `a` being `newvalue` until you eventually hit a memory barrier somewhere else.

When a compiler sees a memory barrier (e.g., in C11, atomic_thread_fence) it will do the right thing and behave as a compiler barrier as well. A mutex contains a memory barrier, and the compiler barrier is usually the function call. Code is either correct or not correct regardless of the `volatile`.

http://lwn.net/Articles/233482/?hc_location=ufi

Also, I don't know if you want Torch7 to require only the C99 standard or the C11 standard, but if you were ok with C11, you could use stdatomic.h, which should be more portable and possibly better performance (see http://en.cppreference.com/w/c/atomic/atomic_fetch_add etc)

http://en.cppreference.com/w/c/atomic